### PR TITLE
TINY-7228: Improve RTC documentation and error messages with customer feedback

### PR DESCRIPTION
**Pull request by TheSpyder**
Related Ticket: TINY-7228

Description of Changes:
* Minor RTC API updates in configuration.
* Added new error message descriptions to RTC troubleshooting.
* New RTC encryption page with detailed descriptions of what encryption means and how it works.
  * The final section here breaks our documentation guidelines, it's very very low level and full of scary cryptography terms, but I would like it to be published _somewhere_ to help security professionals trust that we are doing encryption correctly.
* Moved encryption sections that were incorrectly added to authentication to the new encryption page.

I'm not 100% sure about some of the site tokens I used and the header details of the new page. Please let me know if I need to adjust anything.

Pre-checks:
- [x] Branch prefixed with `feature/` or `hotfix/`
- [x] `_data/nav.yml` has been updated (if applicable)
- [x] Files has been included where required (if applicable)
- [x] ~Files removed have been deleted, not just excluded from the build (if applicable)~
- [x] ~(New product features only) Release Note added~

Review:
- [x] Documentation Team Lead has reviewed
- [x] Product Manager has reviewed

### CODE REVIEW

File: rtc/encryption.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+## Choosing an encryption key
+
+RTC encryption keys can be any unicode string. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`. Even simple keys provide some obfuscation and protection if the key is disclosed; for details on this see [RTC encryption details](#rtcencryptiondetails) below. Simple fixed keys can be a good way to get a test environment up and running but we strongly recommend using secure keys when collaborating on private production data.
```
Comment (ashfordneil) [inline rtc/encryption.md:25]: - Can we change `unicode` to `utf-8` or appropriate throughout these docs?
- I am strongly against suggesting that people use insecure things, and then adding a "but please change it before you go to prod" because people never change it before they go to prod
- "provide some obfuscation and protection if the key is disclosed" sounds like one of the words should have a negative in it

Comment (TheSpyder) [reply rtc/encryption.md:25]: > I am strongly against suggesting that people use insecure things, and then adding a "but please change it before you go to prod" because people never change it before they go to prod

I generally agree, but in this case doing that will raise the barrier to entry too high. RTC is far more complicated to set up than any feature we've ever released before, so we need some way to ease people into it. We've even considered backing down on the signed JWT requirement for test environments but that's not viable on our cloud services.

File: rtc/troubleshooting.md [hunk 2]
```diff
@@ -29,8 +29,36 @@ As a result of this process:
...
+
+### Time-based error messages
+
+These errors include:
+
+* JWT was issued in the future
+* JWT has expired
+* JWT is not yet valid
+
+JWT is a very strict protocol. If the computer that signs the JWT has a clock that is not synchronised with the server clock, and tokens are generated on demand, one of these errors can occur.
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:48]: right, yes, I was doing a lot of protocol wording in this PR. oops.

File: rtc/troubleshooting.md [hunk 3]
```diff
@@ -29,8 +29,36 @@ As a result of this process:
...
 
-### Server reports JWT authentication was issued in the future
+## Common JWT error messages
 
-JWT is a very strict protocol. If the computer that signs the JWT has a clock that is significantly ahead of the server clock, and tokens are signed on demand, this error can occur.
+If the below error descriptions do not immediately solve the problem, try pasting the token into [https://jwt.io/](https://jwt.io/) which will provide more detailed token format debugging.
+
+### Unsupported JWT signature algorithm
+
+RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signatures use separate private and public keys; the public key stored on the {{site.cloudname}} server can only be used to verify signatures, not create new ones. The private key, used to create new signatures, remains private.
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:38]: That seems bad for SEO 🤔

File: rtc/troubleshooting.md [hunk 4]
```diff
@@ -29,8 +29,36 @@ As a result of this process:
...
+* JWT has expired
+* JWT is not yet valid
+
+JWT is a very strict protocol. If the computer that signs the JWT has a clock that is not synchronised with the server clock, and tokens are generated on demand, one of these errors can occur.
 
 To resolve this issue ensure all computer clocks are synchronised using [NTP](https://en.wikipedia.org/wiki/Network_Time_Protocol) or a similar service.
+
+### Unable to retrieve JWK set
+
+A "JWK set" is the technical term for a public key that will be used to validate the JWT signature. It must match the private key used to sign the JWT. This error means a public key could not be found in the {{site.cloudname}} account that matches the API key used to load RTC. For more information on see the [Private/public key pairs for Tiny Cloud services]({{site.baseurl}}/rtc/jwt-authentication/#privatepublickeypairsfortinycloudservices) documentation.
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:54]: A JWK represents a cryptographic key, but yes I shouldn't use the specific wording I did - the JWK set _contains_ the public key is the wording I should have used.

File: rtc/configuration.md [hunk 5]
```diff
@@ -65,10 +65,10 @@ If keys are never rotated this can be ignored. For advice on how to use the key
...
 #### Return fields for `rtc_encryption_provider`
 
-| Field | Type | Description |
-|-------|:----:|-------------|
-| `key` | `string` | Encryption key that is used to locally encrypt operations. This key needs to be the same for all connecting clients on the same session. |
-| `keyHint` | `string` | Key hint to provide to future clients to aid in key selection. If keys are never rotated, it can be a fixed arbitrary value. It is only recorded when the input `keyHint` is `null`. (unicode, max 256 characters) |
+| Field | Type | Required? | Description |
+|-------|:----:|:----:|-------------|
+| `key` | `string` | required | Encryption key that is used to locally encrypt operations. This key needs to be the same for all connecting clients on the same session. |
+| `keyHint` | `string` | optional | Key hint to provide to future clients to aid in key selection. It is only recorded when the input `keyHint` is `null`. (unicode, max 256 characters) |
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/configuration.md:71]: I covered that in the encryption page but good idea, I'll add a note here too.

File: rtc/encryption.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
+---
+layout: default
+title: Encryption information for the RTC protocol
+title_nav: RTC Encryption help
+description: Useful information for understanding how encryption is used with RTC
+keywords: rtc encrypt encryption decrypt decryption key rotate signature
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/encryption.md:6]: Probably not. I haven't written many new pages in these docs.

File: rtc/encryption.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+---
+
+This documentation is in progress. Please contact us with any suggestions you think should be here.
+
+> **Caution**: The advice on this page may not guarantee a secure connection. If data secrecy is important to you please consult a security professional.
+
+## Encryption keys are mandatory
+
+RTC requires an encryption key so {{site.companyname}} can prove to ourselves (and any interested customers) that the RTC server does not store document content in a format that we are able to read. We are very focussed on privacy and end-to-end encryption for this feature.
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/encryption.md:15]: This is copy/pasted from an email I wrote to a customer, happy to take suggestions. It's intended to be a one line summary of [Tim's blog post](https://www.tiny.cloud/blog/real-time-collaboration-end-to-end-encryption/).

File: rtc/encryption.md (hunk 1 above)
Comment (TheSpyder) [reply rtc/encryption.md:25]: Unicode came from the code, and is technically more correct (JavaScript strings are unicode not utf-8). But if we use utf-8 in other TinyMCE docs I'm happy to change it.

I'll see if I can reword the obfuscation piece, it's already up to the third rewrite. Explaining this stuff in simple terms is _hard_.

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:38]: I do that when the link covers a lot more than the topic, but here the link is _specifically_ about the asymmetric signing algorithms we support 🤔

File: rtc/troubleshooting.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:54]: I don't think we need to clarify our use of JWK, as far as customers are concerned they just log their public key in their Tiny Cloud account. We do the rest.

File: rtc/encryption.md (hunk 1 above)
Comment (TheSpyder) [reply rtc/encryption.md]: There are 3 references to `utf-8` in our docs; two in enterprise spelling and one in a config block. There are 8 uses of `unicode` (10 after including RTC).

For now, I'll stick to the more common term we use; perhaps consult Tyler about changing it across our entire documentation site.

I've pushed up a rewrite of this section; I even split off a sub heading "fixed keys are not secure". Hopefully that helps.

**Comment by TheSpyder** [comment]
I am on leave all of next week. Specific questions about encryption should be directed to @tjdett if they can't wait for my return.

This documentation update is not really intended to be production grade. RTC is still a beta. Every page still says "documentation is in progress" or something similar. It still needs to be professional, but I figure we'll have a big correctness audit leading up to the full release.

There's no rush to merge this PR but it does aim to address actual feedback we've received from people trying to integrate RTC. So it would be nice to merge it soonish.

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashfordneil) [reply rtc/troubleshooting.md:38]: Why not `For more information on the asymmetric signing algorithms we support, see [Name of Page — Subheading](...)`?

File: rtc/encryption.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+## Choosing an encryption key
+
+RTC encryption keys can be any unicode string. {{site.companyname}} strongly recommends using secure keys when collaborating on private production data, but for test environments a simple fixed key is still valid. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`.
+
+### Fixed keys are not secure
+
+Even simple keys provide some protection against data disclosure; for details on this see [RTC encryption details](#rtcencryptiondetails) below. Simple fixed keys can be a good way to get started with RTC but if a key is disclosed it can be used to access any document encrypted with that key. Fixed keys mean _every_ document can be decrypted if the key is disclosed.
```
Comment (tylerkelly13) [inline rtc/encryption.md:29]:
```suggestion
Simple, fixed encryption keys provide minimal protection against data disclosure; for details see: [RTC encryption details](#rtcencryptiondetails). Simple fixed keys can be a good way to get started with RTC but if a key is disclosed, it can be used to access any document encrypted with that key. Fixed keys allow _every_ document can be decrypted if the key is disclosed.
```

File: rtc/encryption.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+> **Caution**: The advice on this page may not guarantee a secure connection. If data secrecy is important to you please consult a security professional.
+
+## Encryption keys are mandatory
+
+RTC requires an encryption key so {{site.companyname}} can prove to ourselves (and any interested customers) that the RTC server does not store document content in a format that we are able to read. We are very focussed on privacy and end-to-end encryption for this feature.
+
+## The difference between content encryption and JWT signing
+
+Encryption is a two way process (encryption and decryption) that transforms data into an encrypted form. Modern encryption algorithms produce an output that is indistinguishable from random data without the decryption key. RTC performs symmetric encryption (meaning encryption and decryption use the same key) on all document content, but the key is never sent to the server. As the key only exists on the client endpoints, the term for this is end-to-end encryption.
```
Comment (tylerkelly13) [inline rtc/encryption.md:19]:
```suggestion
Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the user and your server (the client endpoints), and is never sent to the {{site.{{site.cloudname}}}} servers; providing end-to-end encryption.
```

File: rtc/encryption.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+Signing is a one way process where the data is not modified and can still be read by anyone. A signature is transmitted along with the data and can be used to validate the data has not been modified since it was approved by whoever has the signing key. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the Tiny Cloud servers can only be used to verify signatures, not create new ones, so it is quite safe for us to hold or cache without security implications should it be disclosed.
```
Comment (tylerkelly13) [inline rtc/encryption.md:21]:
```suggestion
JWT signing does not modify the data, and the data can be read by anyone (the data is not encrypted). A JWT signature is transmitted along with the data and is used to validate that the data has not been modified since the signature (JWT) was created. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the {{site.cloudname}} servers can only be used to verify signatures, not create new ones. This allows the {{site.cloudname}} server to verify that the data has not been tampered between the end-points, but does not allow {{site.companyname}} to decrypt the data.
```

File: rtc/encryption.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+RTC encryption keys can be any unicode string. {{site.companyname}} strongly recommends using secure keys when collaborating on private production data, but for test environments a simple fixed key is still valid. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`.
+
+### Fixed keys are not secure
+
+Even simple keys provide some protection against data disclosure; for details on this see [RTC encryption details](#rtcencryptiondetails) below. Simple fixed keys can be a good way to get started with RTC but if a key is disclosed it can be used to access any document encrypted with that key. Fixed keys mean _every_ document can be decrypted if the key is disclosed.
+
+## Generating a secure encryption key
+
+Encryption security is a trade off between the complexity of generating a key and the risk of compromise should the key be disclosed to an unknown third party. Here are some suggested ways to generate keys, in descending order of safety:
```
Comment (tylerkelly13) [inline rtc/encryption.md:33]:
```suggestion
Encryption security is a trade off between: the complexity of generating a key, and the risk of compromise should the key be disclosed to an unknown third-party. Methods generate keys include (in descending order of safety):
```

File: rtc/encryption.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+> **Warning**: The key hint is stored by the server in plain text. Do not add secret or sensitive information to the key hint.
+
+The key hint can be a key thumbprint, ID, or other non-sensitive identifier that will help select the key, such as a timestamp. It is only recorded when the protocol requests a new key. Returning a key hint from requests for previously issued keys will have no effect.
+
+## RTC encryption details
+
+> **Note**: This section contains the technical details of encryption used to securely transmit document contents. It is provided for information purposes only; understanding these details is not required to integrate RTC.
+
+The RTC protocol does not use the [provided encryption key]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) to encrypt content directly. Using industry standard cryptography algorithms a unique session key for content encryption is derived from the provided key in a way that makes brute force decryption very difficult.
```
Comment (tylerkelly13) [inline rtc/encryption.md:57]:
```suggestion
The RTC plugin does not use the [provided encryption key]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) to encrypt content directly. The plugin uses industry standard cryptography algorithms to generate a unique session key for content encryption based on the provided key to improve protection against brute force decryption.
```

File: rtc/encryption.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+> **Note**: {{site.companyname}} is in the process of changing RTC encryption to [AES-KW](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey) which does not require an initialization vector or protocol salt. The description below is still correct for the current beta release.
+
+The RTC protocol encryption technique is as follows:
+* 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated with the [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a `Uint8Array` of length 32 bytes.
```
Comment (tylerkelly13) [inline rtc/encryption.md:62]:
```suggestion
* 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated using the [`Crypto.getRandomValues()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a 32 byte long [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array).
```

File: rtc/encryption.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+* The salt data and the provided encryption key are combined using the [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:
+  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations
+  * AES-GCM [derivedKeyAlgorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesKeyGenParams) with length 256
```
Comment (tylerkelly13) [inline rtc/encryption.md:65]:
```suggestion
  * AES-GCM [derivedKeyAlgorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesKeyGenParams) with a length of 256.
```

File: rtc/encryption.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+* The derived key is cached (in memory only) and used with [SubtleCrypto.encrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) and [SubtleCrypto.decrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) as necessary, specifying the [AES-GCM algorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams). For each encrypted message 96 bits of random data are used as the initialization vector, also generated with [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) but this time using `Uint8Array` length 12. Each IV is transmitted alongside the encrypted message.
```
Comment (tylerkelly13) [inline rtc/encryption.md:66]:
```suggestion
* The derived key is cached (in memory only) and used with [SubtleCrypto.encrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) and [SubtleCrypto.decrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) as necessary, specifying the [AES-GCM algorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams). For each encrypted message, 96 bits of random data are used as the initialization vector, also generated with [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) but this time using `Uint8Array` with a length of 12. Each IV is transmitted alongside the encrypted message.
```

Comment (tylerkelly13) [reply rtc/encryption.md:66]: > Each IV is transmitted alongside the encrypted message.

What's `IV`?

File: rtc/encryption.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+> **Note**: {{site.companyname}} is in the process of changing RTC encryption to [AES-KW](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey) which does not require an initialization vector or protocol salt. The description below is still correct for the current beta release.
+
+The RTC protocol encryption technique is as follows:
+* 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated with the [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a `Uint8Array` of length 32 bytes.
+* The salt data and the provided encryption key are combined using the [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:
+  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations
```
Comment (tylerkelly13) [inline rtc/encryption.md:64]:
```suggestion
  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations,
```

File: rtc/encryption.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  * AES-GCM [derivedKeyAlgorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesKeyGenParams) with length 256
+* The derived key is cached (in memory only) and used with [SubtleCrypto.encrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/encrypt) and [SubtleCrypto.decrypt()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/decrypt) as necessary, specifying the [AES-GCM algorithm](https://developer.mozilla.org/en-US/docs/Web/API/AesGcmParams). For each encrypted message 96 bits of random data are used as the initialization vector, also generated with [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) but this time using `Uint8Array` length 12. Each IV is transmitted alongside the encrypted message.
+* The provided key and derived key are never transmitted between clients. Each client will already have the provided key, so only the salt needs to be transmitted for clients to recreate the derived key. The salt alone cannot be used to decrypt the document contents.
```
Comment (tylerkelly13) [inline rtc/encryption.md:67]:
```suggestion
* The provided key and derived key are never transmitted between clients. Each client will already have the provided key, so only the salt needs to be transmitted for clients to recreate the derived key. The salt alone cannot be used to decrypt the document content.
```

File: rtc/encryption.md [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+* When a session is established the provided `keyHint` and protocol-generated salt are sent to the server via a HTTPS websocket connection where they are stored in a database entry linked to the session. When subsequent clients connect to the session, the server sends these values along with other session information so the client can replicate the unique session key using [rtc_encryption_provider]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) and key derivation.
```
Comment (tylerkelly13) [inline rtc/encryption.md:68]:
```suggestion
* When a Real-time collaboration session is established, the provided `keyHint` and protocol-generated salt are sent to the server through a HTTPS websocket connection, where they are stored in a database-entry linked to the session. When subsequent clients connect to the session, the server sends these values along with other session information so the client can replicate the unique session key using [rtc_encryption_provider]({{site.baseurl}}/rtc/configuration#rtc_encryption_provider) and key derivation.
```

File: rtc/encryption.md [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+Our choice of encryption parameters results in an expensive derivation process even on 2021 hardware, but the derivation is only performed once by each client when connecting to a session. {{site.companyname}} is committed to security and privacy of customer data; we chose to favour increased security over derivation performance. Once derivation is complete, encryption and decryption will be hardware accelerated.
```
Comment (tylerkelly13) [inline rtc/encryption.md:70]:
```suggestion
Our choice of encryption parameters results in an expensive derivation process, even on modern hardware, but the derivation is only performed once by each client when connecting to a session. {{site.companyname}} is committed to the security and privacy of customer data, and {{site.companyname}} has prioritized increased security over derivation performance. Once derivation is complete, encryption and decryption will be hardware accelerated.
```

File: rtc/encryption.md [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+> **Note**: {{site.companyname}} is in the process of changing RTC encryption to [AES-KW](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/wrapKey) which does not require an initialization vector or protocol salt. The description below is still correct for the current beta release.
+
+The RTC protocol encryption technique is as follows:
+* 256 bits of random data are generated as the salt for each session. As described above, each document ID used for collaboration may have multiple sessions. The salt is generated with the [Crypto.getRandomValues()](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a `Uint8Array` of length 32 bytes.
+* The salt data and the provided encryption key are combined using the [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:
```
Comment (tylerkelly13) [inline rtc/encryption.md:63]: Regarding the term "key derivation function" and "derivation", we need to either:
- link to a reliable source
- provide a description
- avoid using
I know this is a technical section, but that also means it's easy to get lost.

**Review by tylerkelly13 — COMMENTED** [review]
Looks good, needs some reworking

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tylerkelly13) [reply rtc/troubleshooting.md:38]: `more` is unnecessary, `we` is unclear, and that sentence is written in passive voice: https://www.tiny.cloud/docs/configure/contributing-docs/#useactivevoicenotpassivevoice

File: _data/nav.yml [hunk 21]
```diff
@@ -696,7 +702,7 @@
       pages:
       - url: "#RtcClientConnected"
       - url: "#RtcClientDisconnected"
-    - url: "rtc-troubleshooting"
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply _data/nav.yml:699]: I thought this was more consistent. None of the other files in RTC have an `rtc-` prefix, why is troubleshooting special?

File: rtc/encryption.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/encryption.md:15]: I might make some tweaks for correctness, but this is good. Thanks!

File: rtc/encryption.md (hunk 9 above)
Comment (TheSpyder) [reply rtc/encryption.md:19]: again, I'll correct after committing your suggestions - the key isn't provided _to_ the user, technically, or if it is it's provided by the integrating server. Is `{{site.{{site.cloudname}}}}` valid? I've never seen that before 🧐

File: rtc/encryption.md (hunk 10 above)
Comment (TheSpyder) [reply rtc/encryption.md:21]: I was deliberately trying to describe signing in general, the process is not unique to JWT, but I guess for anyone unfamiliar with it the difference is meaningless.

File: rtc/encryption.md [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+Signing is a one way process where the data is not modified and can still be read by anyone. A signature is transmitted along with the data and can be used to validate the data has not been modified since it was approved by whoever has the signing key. RTC requires an [asymmetric signing algorithm]({{site.baseurl}}/rtc/jwt-authentication/#supportedalgorithms) for JWT identity tokens. Asymmetric signing uses a public/private key pair: signatures are created with the private key and verified using the public key. The public key stored on the Tiny Cloud servers can only be used to verify signatures, not create new ones, so it is quite safe for us to hold or cache without security implications should it be disclosed.
+
+## Choosing an encryption key
+
+RTC encryption keys can be any unicode string. {{site.companyname}} strongly recommends using secure keys when collaborating on private production data, but for test environments a simple fixed key is still valid. Our demo application for example uses the simple fixed key `"not a very secret 🔑"`.
```
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/encryption.md:25]: The demo application hasn't been released yet, so I can't link to it. This will do for now.

File: rtc/encryption.md (hunk 12 above)
Comment (TheSpyder) [reply rtc/encryption.md:57]: I appreciate that you're trying to make things easier to follow, but for the target audience here the word "derived" is _very_ important to include. I'll put it back after the suggestions are committed.

File: rtc/encryption.md (hunk 20 above)
Comment (TheSpyder) [reply rtc/encryption.md:63]: wikipedia isn't reliable?

Perhaps this entire section needs to be shifted to a dedicated "security professionals, here is how we can prove to you we know what we're talking about" page instead of tacking it onto the end of the general integrator-targeted encryption page.

File: rtc/encryption.md (hunk 15 above)
Comment (TheSpyder) [reply rtc/encryption.md:66]: Initialization Vector. I guess I can repeat that rather than using the acronym.

File: rtc/encryption.md (hunk 18 above)
Comment (TheSpyder) [reply rtc/encryption.md:68]: `third-party` doesn't have a hyphen, but `database-entry` does? 🧐

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:38]: I did this on [the other page too](https://github.com/tinymce/tinymce-docs/pull/1916/files?file-filters%5B%5D=.md#diff-2bcdccd2556c30529dbafd69d004d4b74d1f09e042a9daf649333c4426e47d15R21), and it wasn't picked up. I think linking mid-sentence is better UX.

[the remark this answers is not in the document]
Comment (tylerkelly13) [reply rtc/troubleshooting.md:38]: not really. I did see this in the very technical part but ignored it because it's long with too many links to pull out. screen readers will get mid-sentence, then read out a URL, so bad UX

File: _data/nav.yml (hunk 21 above)
[the remark this answers is not in the document]
Comment (tylerkelly13) [reply _data/nav.yml:699]: it's rtc specific, not much chance for reuse. there's also a chance of other troubleshooting pages popping up in the future.

File: rtc/encryption.md (hunk 18 above)
Comment (tylerkelly13) [reply rtc/encryption.md]: no, it does. if I missed it, please fix third-party refs (hyphen required)
Added the hyphen here for readability. those three words in a row are hard to parse, so the hyphen was added to show association between to the two most-related words.

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:38]: If we used relative links, it would probably be fine. The way we do it is called out as bad as well.
https://usability.yale.edu/web-accessibility/articles/links#link-text

File: rtc/encryption.md (hunk 18 above)
Comment (TheSpyder) [reply rtc/encryption.md]: Ah my apologies I misremembered the earlier change - you _added_ one, you didn't remove one. There's a lot to keep track of.

File: rtc/encryption.md (hunk 20 above)
Comment (tylerkelly13) [reply rtc/encryption.md:72]: > wikipedia isn't reliable?

no. it's not.

Regarding your second point, maybe? This is something we could try address now or before the RTC 1.0 release.

Comment (TheSpyder) [reply rtc/encryption.md:72]: I could link to the IETF RFC, but that's _completely_ unintelligible. Wikipedia at least gives an average user a chance if they try to follow this.

I think moving it to a separate page can be a pre-1.0 task, rather than trying to figure it out now.

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (tylerkelly13) [reply rtc/troubleshooting.md:38]: yeah, that's about people doing this: 
image
which falls under my first point here: https://www.tiny.cloud/docs/configure/contributing-docs/#linksandcross-references

File: rtc/encryption.md (hunk 20 above)
Comment (TheSpyder) [reply rtc/encryption.md:72]: Google for PBKDF2 also brings up https://cryptobook.nakov.com/mac-and-key-derivation/pbkdf2 but that's someone's personal site by the looks of it. A professional someone, but I don't know if it will meet your standard for reliability.

File: rtc/troubleshooting.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheSpyder) [reply rtc/troubleshooting.md:38]: I think it applies to "see: <link text>" as well, but ok.

File: rtc/encryption.md (hunk 20 above)
Comment (TheSpyder) [reply rtc/encryption.md]: I went with the IETF. Google will bring up wikipedia, we don't need to link to it.

**Review by JamesToohey — APPROVED** [review]

File: rtc/encryption.md [hunk 23]
```diff
@@ -1,36 +1,45 @@
...
+- To decrypt content clients receive from the {{site.cloudname}} server.
+
+{{site.productname}} clients use the encryption key but do not transmit it to the {{site.cloudname}} servers. {{site.companyname}} **will never have the key to decrypt document content**.
 
 ## The difference between content encryption and JWT signing
 
-Encryption is a two way process (encryption and decryption) that transforms data into an encrypted form. Modern encryption algorithms produce an output that is indistinguishable from random data without the decryption key. RTC performs symmetric encryption (meaning encryption and decryption use the same key) on all document content, but the key is never sent to the server. As the key only exists on the client endpoints, the term for this is end-to-end encryption.
+Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the client by your server, and is never sent to the {{site.cloudname}} servers; that is end-to-end encryption.
```
Comment (tylerkelly13) [inline rtc/encryption.md:26]:
```suggestion
Encrypted data is indistinguishable from random data without the decryption key. The RTC plugin performs client-side symmetric encryption (using the same key for encryption and decryption) on the editor content. The encryption key is provided to the client by your server, and is never sent to the {{site.cloudname}} servers; providing end-to-end encryption.
```

File: rtc/encryption.md [hunk 24]
```diff
@@ -52,19 +61,20 @@ The key hint can be a key thumbprint, ID, or other non-sensitive identifier that
...
\ No newline at end of file
+* As described above, each document ID used for collaboration may have multiple sessions. To ensure each session has a unique key, 256 bits of random data are generated as the salt for each session. The salt is generated using the [`Crypto.getRandomValues()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues) browser API and a 32 byte [`Uint8Array`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array).
+* The salt data and the provided encryption key are combined using a [PBKDF2](https://tools.ietf.org/html/rfc2898#section-5.2) key derivation function. Derivation is performed by the [SubtleCrypto.deriveKey()](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey) browser API, using the following parameters:
+  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10000 iterations,
```
Comment (tylerkelly13) [inline rtc/encryption.md:73]:
```suggestion
  * PBKDF2 [algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Pbkdf2Params), with SHA-512 digest algorithm and 10,000 iterations,
```

**Review by tylerkelly13 — APPROVED** [review]
Couple minor suggestions, otherwise LGTM

**Review by Afraithe — APPROVED** [review]
I think this clarifies a lot of the issues ppl are running into with the setup, good to go!
